ffi/memorymanager.cpp

```diff
@@ -0,0 +1,340 @@
+    return allocateSection(LlvmliteMemoryManager::AllocationPurpose::Code, Size,
+                           Alignment);
+}
+
+uint8_t *LlvmliteMemoryManager::allocateSection(
+    LlvmliteMemoryManager::AllocationPurpose Purpose, uintptr_t Size,
+    unsigned Alignment) {
+    if (!Alignment)
+        Alignment = 16;
+
+    LLVM_DEBUG(dbgs() << "Requested size / alignment: " << format_hex(Size, 2, true) << " / " << Alignment << "\n");
+
+    assert(!(Alignment & (Alignment - 1)) &&
+           "Alignment must be a power of two.");
+
+    uintptr_t RequiredSize =
+        Alignment * ((Size + Alignment - 1) / Alignment + 1);
+    uintptr_t Addr = 0;
+
+    LLVM_DEBUG(dbgs() << "Allocating " << format_hex(RequiredSize, 2, true)
+                      << " bytes for ");
+
+    MemoryGroup &MemGroup = [&]() -> MemoryGroup & {
+        switch (Purpose) {
+        case AllocationPurpose::Code:
+            LLVM_DEBUG(dbgs() << "CodeMem at ");
+            return CodeMem;
+        case AllocationPurpose::ROData:
+            LLVM_DEBUG(dbgs() << "RODataMem at ");
+            return RODataMem;
+        case AllocationPurpose::RWData:
+            LLVM_DEBUG(dbgs() << "RWDataMem at ");
+            return RWDataMem;
+        }
+        llvm_unreachable("Unknown LlvmliteMemoryManager::AllocationPurpose");
+    }();
+
+    // Look in the list of free memory regions and use a block there if one
+    // is available.
+    for (FreeMemBlock &FreeMB : MemGroup.FreeMem) {
+        if (FreeMB.Free.allocatedSize() >= RequiredSize) {
+            Addr = (uintptr_t)FreeMB.Free.base();
+            uintptr_t EndOfBlock = Addr + FreeMB.Free.allocatedSize();
+            // Align the address.
+            Addr = (Addr + Alignment - 1) & ~(uintptr_t)(Alignment - 1);
+
+            if (FreeMB.PendingPrefixIndex == (unsigned)-1) {
+                // The part of the block we're giving out to the user is now
+                // pending
+                MemGroup.PendingMem.push_back(
+                    sys::MemoryBlock((void *)Addr, Size));
+
+                // Remember this pending block, such that future allocations can
+                // just modify it rather than creating a new one
+                FreeMB.PendingPrefixIndex = MemGroup.PendingMem.size() - 1;
+            } else {
+                sys::MemoryBlock &PendingMB =
+                    MemGroup.PendingMem[FreeMB.PendingPrefixIndex];
+                PendingMB =
+                    sys::MemoryBlock(PendingMB.base(),
+                                     Addr + Size - (uintptr_t)PendingMB.base());
+            }
+
+            // Remember how much free space is now left in this block
+            FreeMB.Free = sys::MemoryBlock((void *)(Addr + Size),
+                                           EndOfBlock - Addr - Size);
+            LLVM_DEBUG(dbgs() << format_hex(Addr, 18, true) << "\n");
+            return (uint8_t *)Addr;
+        }
+    }
+
+    assert(false && "All memory must be pre-allocated");
+}
```

Comment:
Do we need to return a nullptr here for builds that don't have asserts?

Reply:
Yes - on my Linux AArch64 system I was getting a hang in the `test_nanquantile_basic` test that hits this assert on my M2 system, and I'd guess that might be the reason why - a null pointer should allow it to crash more quickly :-)